storage/storage_image.go

```diff
@@ -0,0 +1,569 @@
+			return err
+		}
+		logrus.Debugf("saved image metadata %q", string(metadata))
+	}
+	return nil
+}
+
+func (s *storageImageDestination) SupportedManifestMIMETypes() []string {
+	return nil
+}
+
+func (s *storageImageDestination) PutManifest(manifest []byte) error {
+	s.Manifest = make([]byte, len(manifest))
+	copy(s.Manifest, manifest)
+	return nil
+}
+
+// SupportsSignatures returns an error if we can't expect GetSignatures() to
+// return data that was previously supplied to PutSignatures().
+func (s *storageImageDestination) SupportsSignatures() error {
+	return nil
+}
+
+func (s *storageImageDestination) PutSignatures(signatures [][]byte) error {
+	sizes := []int{}
+	sigblob := []byte{}
+	for _, sig := range signatures {
+		sizes = append(sizes, len(sig))
+		newblob := make([]byte, len(sigblob)+len(sig))
+		copy(newblob, sigblob)
+		copy(newblob[len(sigblob):], sig)
+		sigblob = newblob
+	}
+	s.Signatures = sigblob
+	s.SignatureSizes = sizes
+	return nil
+}
+
+func (s *storageImageSource) GetBlob(digest string) (rc io.ReadCloser, n int64, err error) {
+	rc, n, _, err = s.GetBlobAndLayerID(digest)
+	return rc, n, err
+}
+
+func (s *storageImageSource) GetBlobAndLayerID(digest string) (rc io.ReadCloser, n int64, layerID string, err error) {
+	err = ddigest.Digest(digest).Validate()
+	if err != nil {
+		return nil, -1, "", err
+	}
+	if layerList, ok := s.Layers[digest]; !ok || len(layerList) < 1 {
+		b, err := s.imageRef.transport.store.GetImageBigData(s.ID, digest)
+		if err != nil {
+			return nil, -1, "", err
+		}
+		r := bytes.NewReader(b)
+		logrus.Debugf("exporting opaque data as blob %q", digest)
+		return ioutil.NopCloser(r), int64(r.Len()), "", nil
+	}
+	// If the blob was "put" more than once, we have multiple layer IDs
+	// which should all produce the same diff.  For the sake of tests that
+	// want to make sure we created different layers each time the blob was
+	// "put", though, cycle through the layers.
+	layerList := s.Layers[digest]
+	position, ok := s.LayerPosition[digest]
+	if !ok {
+		position = 0
+	}
+	s.LayerPosition[digest] = (position + 1) % len(layerList)
+	logrus.Debugf("exporting filesystem layer %q for blob %q", layerList[position], digest)
+	rc, n, err = s.diffLayer(layerList[position])
+	return rc, n, layerList[position], err
+}
+
+func (s *storageImageSource) diffLayer(layerID string) (rc io.ReadCloser, n int64, err error) {
+	layer, err := s.imageRef.transport.store.GetLayer(layerID)
+	if err != nil {
+		return nil, -1, err
+	}
+	layerMeta := storageLayerMetadata{
+		CompressedSize: -1,
+	}
+	if layer.Metadata != "" {
+		if err := json.Unmarshal([]byte(layer.Metadata), &layerMeta); err != nil {
+			logrus.Errorf("error decoding metadata for layer %q: %v", layerID, err)
+			return nil, -1, err
+		}
+	}
+	if layerMeta.CompressedSize <= 0 {
+		n, err = s.imageRef.transport.store.DiffSize("", layer.ID)
+		if err != nil {
+			return nil, -1, err
+		}
+	} else {
+		n = layerMeta.CompressedSize
```

Comment:
A layer blob isn't stored in its original form, but we record the original and uncompressed size when extracting it.  We've inherited an implicit assumption that gzip-compressed blobs are compressed using `compress/gzip` at the DefaultCompression level, so for blobs that were passed in compressed, GetBlob() hands back the compressed size that was previously recorded, compresses the reconstructed blob on the way out, and hopes for the best.  We could discard the compression instead, but that would break things elsewhere.

I'll change the fallback to default to not recomputing the blob's size if it isn't already known.